### ISSUE TITLE
SubjectGroup selection bug - incorrect URLs

### ISSUE THREAD

**Issue by shaunanoordin** [zooniverse/panoptes#3549]
## API bug

The URLs returned by the SubjectGroup API are in an incorrect format.

- example of an URL from API: https://panoptes-uploads.zooniverse.org/staging/subject_location/b71027f2-2416-4d8f-aa42-316791ca0d8a.png
- CORRECT URL: https://panoptes-uploads-staging.zooniverse.org/subject_location/b71027f2-2416-4d8f-aa42-316791ca0d8a.png

Related: #3530 - the initial addition of the SubjectGroup API end points

### Testing

Setup
- Project: [Survos Survey 2020 (staging)](https://master.pfe-preview.zooniverse.org/lab/1900)
- Workflow: [3412](https://master.pfe-preview.zooniverse.org/lab/1900/workflows/3412) - workflow.configuration set up to use the SubjectGroupViewer and the Survos API. Whitelisted to use the Survos API.
- Subject Set: Subject Set 4780, which has 6 Subjects, each having 25 images

Action:
- pinged https://panoptes-staging.zooniverse.org/api/subjects/grouped?workflow_id=3412&num_rows=5&num_columns=5 via Postman

Expected:
- A pseudo-Subject that consists of 25 images. Resource looks the same as any other Subject, as the Front End is concerned.

Result:
- A valid pseudo-Subject is received, but the URLs are all invalid.

Example:
```
{
    "subjects": [
        {
            "id": "134898",
            "metadata": {
                "#subject_group_id": 6,
                "#group_subject_ids": "85110-85111-85112-85114-85113-85109"
            },
            "locations": [
                {
                    "image/png": "https://panoptes-uploads.zooniverse.org/staging/subject_location/b71027f2-2416-4d8f-aa42-316791ca0d8a.png"
                },
                {
                    "image/png": "https://panoptes-uploads.zooniverse.org/staging/subject_location/a25dc6fa-e056-4e82-a480-d541ebb3ca8e.png"
                },
                {
                    "image/png": "https://panoptes-uploads.zooniverse.org/staging/subject_location/8ed9d95f-ac06-4868-8323-c9c29658bfb5.png"
                },
                ...total of 150 URLs...
            ],
            "zooniverse_id": null,
            "created_at": "2021-01-11T15:58:14.786Z",
            "updated_at": "2021-01-11T15:58:14.786Z",
            "href": "/subjects/134898",
            "selected_at": "2021-01-11T15:58:15.206Z",
            "retired": false,
            "already_seen": false,
            "finished_workflow": false,
            "user_has_finished_workflow": false,
            "favorite": false,
            "selection_state": "normal"
        }
    ],
    "meta": ...
}
```

Side note (irrelevant to the issue, but noting for posterity):
- I receive 150 image URls instead of 25... which sorta makes sense, because the API finds only 6 Subjects out of 25, pulls them all anyway, and it turns out those 6 Subs have 25 images each (6*25 = 150). This is fine, because Subject Set 4780 isn't using the "correct" setup of 500 Subjects with one image each.

### Status

Low priority bug, because the SubjectGroup feature is used only for Survos, and the front end can create an adaptor/transformer to make that URL work.

Tagging @camallen

**Comment by camallen** [comment]
I've added a new redirect rule on the `panoptes-uploads.zooniverse.org` frontdoor CDN to redirect any `/staging/*` path urls to the `panoptes-uploads-staging.zooniverse.org` domain. These images should now load correctly in the UI. 
``` bash
$ curl -I https://panoptes-uploads.zooniverse.org/staging/subject_location/b71027f2-2416-4d8f-aa42-316791ca0d8a.png
HTTP/2 302 
content-length: 0
location: https://panoptes-uploads-staging.zooniverse.org/staging/subject_location/b71027f2-2416-4d8f-aa42-316791ca0d8a.png
server: Microsoft-IIS/10.0
x-azure-ref: 0dF4AYAAAAACquDHif9PaQpqqDROIItpKTE9OMjFFREdFMDEwOAAyMDgxYjI0OS00MTYyLTQ5YmUtOWMwYy1mMTk1YzBkODljODY=
date: Thu, 14 Jan 2021 15:08:36 GMT
```

**Comment by camallen** [comment]
> Side note (irrelevant to the issue, but noting for posterity):

    I receive 150 image URls instead of 25... which sorta makes sense, because the API finds only 6 Subjects out of 25, pulls them all anyway, and it turns out those 6 Subs have 25 images each (6*25 = 150). This is fine, because Subject Set 4780 isn't using the "correct" setup of 500 Subjects with one image each.

This is a good point - i think i'll add a check on the grouped subject creator to ensure that there is only 1 image media resource for each 'group' subject, otherwise, the resulting media resources can explode and possibly break the UI.

**Comment by camallen** [comment]
I think this is closed by the CDN changes and the PR fixes in #3554 so i'm going to close for now. Feel free to re-open / comment on this issue if it's not fixed and i'll re-open.

**Comment by shaunanoordin** [comment]
## Bug Un-Squashed

I'm reopening this issue, sorry! Following #3554 I've done some tests I've found the Subject Group API endpoint introducing a _slightly different variation_ of the erroneous URL.

Action:
- pinged a GET request to `https://panoptes-staging.zooniverse.org/api/subjects/grouped?workflow_id=3412&num_rows=5&num_columns=5` via Postman

Result:
```
{
    "subjects": [
        {
            "id": "134955",
            "metadata": {
                "#subject_group_id": 63,
                "#group_subject_ids": "134711-134822-134706-134861-134855-134603-134845-134691-134820-134617-134707-134630-134866-134769-134893-134658-134612-134754-134869-134885-134743-134745-134598-134749-134686"
            },
            "locations": [
                {
                    "image/png": "https://subject_location/7d5b2efb-31c7-456d-86cd-e27a0d77621d.png"
                },
                {
                    "image/png": "https://subject_location/0736d6b9-b3d1-41d5-a192-c7af6657aaed.png"
                },
                {
                    "image/png": "https://subject_location/f734e36b-56cc-4d0c-9fab-d68aa0462c29.png"
                },
                {
                    "image/png": "https://subject_location/857a8e3f-5aae-4f87-9987-c2cc140e28ba.png"
                },
                {
                    "image/png": "https://subject_location/8dcd6d18-6c16-4dde-8aea-ac278bab3863.png"
                },
                {
                    "image/png": "https://subject_location/f6fe6a5a-c850-49a6-8c9d-65a86f1d5fb9.png"
                },
                {
                    "image/png": "https://subject_location/3630fc62-e96e-4af5-bfd8-3a5fcd3d3229.png"
                },
                {
                    "image/png": "https://subject_location/32304289-3e1a-402e-8de0-35467d61acc2.png"
                },
                {
                    "image/png": "https://subject_location/9ec81da0-4511-480e-90c9-e79396c14be3.png"
                },
                {
                    "image/png": "https://subject_location/61f6abb8-f1cd-4455-b1c5-520f228e17ed.png"
                },
                {
                    "image/png": "https://subject_location/67f315ab-8219-42a7-98cc-694156d585ca.png"
                },
                {
                    "image/png": "https://subject_location/e4597fba-d3bf-4484-af53-6ced7f7ebef5.png"
                },
                {
                    "image/png": "https://subject_location/440ee50a-4365-4d5e-9182-38c0a6bb9613.png"
                },
                {
                    "image/png": "https://subject_location/f3c0353a-ac86-4bc7-992f-861ca7812c31.png"
                },
                {
                    "image/png": "https://subject_location/e2c44338-7bce-45c0-8a64-c33f3682f7e8.png"
                },
                {
                    "image/png": "https://subject_location/03829109-cd1d-44a5-96d8-19a964434440.png"
                },
                {
                    "image/png": "https://subject_location/2fb859de-024b-4146-a0ef-888f8235e543.png"
                },
                {
                    "image/png": "https://subject_location/421f6d5e-eedb-4c86-87e1-05b97e9fc109.png"
                },
                {
                    "image/png": "https://subject_location/b90e3324-902a-456c-baf9-80284c4d7e0f.png"
                },
                {
                    "image/png": "https://subject_location/d708f076-db21-478f-b0bd-335590941b2b.png"
                },
                {
                    "image/png": "https://subject_location/7f61eed7-93e5-448f-b352-6e2ab0acf922.png"
                },
                {
                    "image/png": "https://subject_location/aa5d1029-1bc8-4f9f-96aa-7cad1e8da018.png"
                },
                {
                    "image/png": "https://subject_location/50ed0cbc-8a4d-45dd-bb59-9b651e8e5c2c.png"
                },
                {
                    "image/png": "https://subject_location/d6ebdf83-60c5-4605-8fe2-cc5d00e8cae7.png"
                },
                {
                    "image/png": "https://subject_location/46ad27cb-128a-4c41-957e-ac474f19d219.png"
                }
            ],
            "zooniverse_id": null,
            "created_at": "2021-01-25T18:08:34.516Z",
            "updated_at": "2021-01-25T18:08:34.516Z",
            "href": "/subjects/134955",
            "selected_at": "2021-01-25T18:08:35.101Z",
            "retired": false,
            "already_seen": false,
            "finished_workflow": false,
            "user_has_finished_workflow": false,
            "favorite": false,
            "selection_state": "normal"
        },
        ...subject 64...,
        ...subject 65...
    ],
    "meta": {
        ...
    }
}
```

Analysis:
- as you can see, the URLs from the API have gone rather bonkers: `"https://subject_location/7d5b2efb-31c7-456d-86cd-e27a0d77621d.png"`
- correct URL should of course be `"https://panoptes-uploads-staging.zooniverse.org/subject_location/7d5b2efb-31c7-456d-86cd-e27a0d77621d.png"`

### Status

I'm genuinely baffled as to what's going on here. I've taken a look at 3554's code and I don't see what could possible cause the `panoptes-uploads-staging.zooniverse.org/` to be stripped out of the image URLs. Is there a missing ENV variable?

Good news is, while this is still an important bug to fix, it remains a low urgency issue at the moment since the front-end has a temporary (simple yet effective) URL replacement workaround for continued Survos development.

**Comment by camallen** [comment]
this was fixed by #3554
